hw_11/task_02.ts

```diff
@@ -0,0 +1,167 @@
+
+console.log("-------------------task_2.1-------------------")
+type TGrade = "junior" | "middle" | "senior" | "lead";
+
+interface ItEmployee {
+    name: string;
+    surname: string;
+    readonly salary: number;
+    grade: TGrade; //значение из типа TGrade
+    occupation: OCCUPATION;
+    address?: IAddress; //необязательное поле, значение из интерфейса 
+    projectNames: string[];
+}
+
+enum OCCUPATION {
+    DEVELOPER = "Developer",
+    QA = "QA Engineer",
+    DESINGER = "UI/UX Desinger",
+}
+
+interface IAddress {
+    country: string;
+    street: string;
+    house: number;
+    flat: number;
+}
+interface IEmployee {
+    name: string;
+    surname: string;
+    readonly salary: number;
+    address?: IAddress;
+}
+
+function isItEmployee(employee: ItEmployee | IEmployee): employee is ItEmployee {
+    return ('occupation' in employee);
+}
+
+function getEmployeeInfo(employee: ItEmployee | IEmployee) { //объект будет создан на основе одного из интерфейсов
+    if (isItEmployee(employee)) {
+        console.log(`IT-Employee ${employee.name} ${employee.surname}, 
+            salary - ${employee.salary}$, 
+            address - ${employee.address?.country}, ${employee.address?.street}, ${employee.address?.house} ap. ${employee.address?.flat}
+            grade - ${employee.grade},  
+            occupation - ${employee.occupation},  
+            projects - ${employee.projectNames}.`)
+    } else
+        console.log(`Employee - ${employee.name} ${employee.surname}, 
+        salary - ${employee.salary}$, 
+        address - ${employee.address?.country}, ${employee.address?.street}, ${employee.address?.house} ap. ${employee.address?.flat}`);
+
+}
+
+const ItEmployee: ItEmployee = {
+    name: 'Elena',
+    surname: 'Al-Koli',
+    salary: 2500,
+    grade: 'junior',
+    occupation: OCCUPATION.QA,
+    projectNames: ['3dGame', 'MerchStore'],
+    address: { country: "Russia", street: "Nevskiy prospect", house: 32, flat: 1 },
+};
+
+const Imployee: IEmployee = {
+    name: 'Elena',
+    surname: 'Al-Koli',
+    salary: 2500,
+    address: { country: "Russia", street: "Nevskiy prospect", house: 32, flat: 1 },
+};
+
+getEmployeeInfo(ItEmployee);
+getEmployeeInfo(Imployee);
+
+// 2. Создайте функцию, которая подсчитает, сколько в объекте значений каждого типа.
+//   Принимает на вход объект или массив таких же объектов, у которого ключ всегда string, а значение - string, number, boolean.
+//   Возвращает же - объект с ключами string, number, boolean и количеством таких значений в объекте или в сумме у всех объектов в массиве.
+console.log("-------------------task_2.2-------------------")
+
+type TObjects = { [key: string]: string | number | boolean }; //объект или массив объектов
+
+interface IValueTypes {
+  string: number;
+  number: number;
+  boolean: number;
+}
+
+function CountValueTypes(input: TObjects | TObjects[]): IValueTypes {
```

Comment:
According to the book "Javascript: the good parts", you should only capitalise the first character of the name of a function when you need to construct the object by "new" keyword.

This is called "the Constructor Invocation Pattern", a way to inherits
```suggestion
function countValueTypes(input: TObjects | TObjects[]): IValueTypes {
```